Introduce `tx_graph::Update` and simplify `TxGraph` update logic

evanlinjin opened the pull request:
Part of #1543
Closes #1550

### Description

Instead of updating a `TxGraph` with another `TxGraph` in `.apply_update()`, we introduce `tx_graph::Update`. `tx_graph::Update` is a simple data object. This is the first step of #1543. This also makes it slightly less expensive to create an update.

Additionally, we simplify the update logic of `TxGraph` by containing most of the update logic in `.insert_{}`-esc methods. `.apply_update` and `.apply_changeset` will call `.insert_{}` methods internally and is greatly simplified. Thus, we also get rid of `.determine_changeset`.

We change `.apply_update` methods of `TxGraph`,`IndexedTxGraph` and `Wallet` to implicitly set `seen_at` to the current timestamp for unconfirmed update transactions. This makes these methods only available when the `"std"` cargo feature is enabled. For a non-std environment, we introduce `.apply_update_at` (as shown below).

```rust
pub fn apply_update_at(&mut self, update: Update<A>, seen_at: Option<u64>) -> ChangeSet<A> {
    todo!()
}
```

During this process, I've fixed some tests that didn't make much sense.

### Notes to the reviewers

There is a slight scope-creep, but they all address the update logic of `TxGraph` and structures that build on top.

### Changelog notice

* Changed update API (`.apply_update`) of `TxGraph` to take in a simple data object (`tx_graph::Update`).
* Changed `.apply_update` methods of `TxGraph`, `IndexedTxGraph` and `Wallet` to implicitly set the `seen_at` to the current timestamp for unconfirmed update transactions (this making them depend on the "std" cargo feature).
* Add `.apply_update_at` which is the no-std version of `.apply_update`.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

evanlinjin committed:
feat(chain): get rid of `TxGraph::determine_changeset`

Contain most of the insertion logic in `.insert_{}` methods, thus
simplifying `.apply_{}` methods. We can also get rid of
`.determine_changeset`.

evanlinjin committed:
feat!: introduce `tx_graph::Update`

Instead of updating a `TxGraph` with a `TxGraph`, we introduce a
dedicated data object (`tx_graph::Update`). This brings us closer to
completing #1543.

Co-authored-by: Wei Chen <wzc110@gmail.com>

evanlinjin committed:
feat(chain)!: `TxGraph::apply_update` auto-adds `seen_at` for unanchored

Change `apply_update` to use the current timestamp as `seen_at` for
unanchored transactions of the update. This makes `apply_update` only
avaliable with the "std" feature.

Introduce `apply_update_at` which includes an optional `seen_at` input.
This is the no-std version of `apply_update`.

Also update docs.

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -640,125 +691,70 @@ impl<A: Clone + Ord> TxGraph<A> {
-    pub fn apply_update(&mut self, update: TxGraph<A>) -> ChangeSet<A> {
-        let changeset = self.determine_changeset(update);
-        self.apply_changeset(changeset.clone());
+    #[cfg(feature = "std")]
```
```suggestion
    #[cfg(feature = "std")]
    #[cfg_attr(docsrs, doc(cfg(feature = "std")))]
```

Explain what's going on here internally.

LLFourn commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -385,7 +310,7 @@ fn insert_txout_does_not_displace_tx() {
         OutPoint {
             txid: tx.compute_txid(),
-            vout: 0,
+            vout: 1,
```
looks like a mistaken change.

LLFourn replied:
oh no it's right.

LLFourn commented on crates/esplora/src/blocking_ext.rs:
```diff
@@ -339,59 +354,48 @@ fn fetch_txs_with_spks<I: IntoIterator<Item = ScriptBuf>>(
                 std::thread::spawn(move || {
-                    if tx_already_exists {
-                        client
-                            .get_tx_status(&txid)
```
It looks like we were calling `get_tx_status` if we already had it but this seems like a bug so you removed it?

evanlinjin replied:
Yes. I wouldn't call it a bug. It's just that esplora always returns a tx with a status. If we already have a tx, that means we already have the status. No need to getting the status here again

LLFourn commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -88,57 +88,21 @@ fn insert_txouts() {
-                    txouts: [(*outpoint, txout.clone())].into(),
-                    ..Default::default()
-                }
-            );
```
note sure if it's worth it but you could keep all the code here the same and then just turn the `TxGraph` into an `Update` at the end. Might as well to test the `From` impl but up to you.

evanlinjin replied:
I think it's better to add separate tests for those!

evanlinjin replied:
Tests added.

evanlinjin committed:
revert(chain)!: rm `TxGraph::update_last_seen_unconfirmed`

This is no longer needed as `TxGraph::apply_update` now automatically
adds `seen_at` timestamps for unanchored transactions.

evanlinjin committed:
docs(chain): use `doc_cfg` feature

evanlinjin committed:
test(chain): `TxGraph` and `tx_graph::Update` conversion tests

notmandatory reviewed (approved):
ACK ccb8c796b259ae348cb8b8ef35b8c60e5621ccb5

Looks good and I can see how this will help decouple the blockchain clients from the wallet once `tx_graph:Update` is moved to a new crate. 

I ran all the examples also as a little smoke test and they worked fine.

LLFourn reviewed (approved):
ACK ccb8c796b259ae348cb8b8ef35b8c60e5621ccb5
